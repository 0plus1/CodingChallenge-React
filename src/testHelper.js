import React from 'react';
import MemoryRouter from 'react-router/MemoryRouter';
import { mount, render, shallow } from 'enzyme';

const mountWithRouter = node => mount(<MemoryRouter>{node}</MemoryRouter>);
const renderWithRouter = node => render(<MemoryRouter>{node}</MemoryRouter>);
const shallowWithRouter = node => shallow(<MemoryRouter>{node}</MemoryRouter>);

export {
  mountWithRouter,
  renderWithRouter,
  shallowWithRouter,
};
