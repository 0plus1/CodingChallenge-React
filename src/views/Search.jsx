import React from 'react';
import PropTypes from 'prop-types';
import Layout from './Layout';
import BookResults from '../components/BookResults';
import { useEffect } from 'react';
import { connect } from 'react-redux';
import { loadBooks } from '../actions/bookActions';
import Loading from './Loading';

const Search = ({ books, dispatch }) => {
  // if the route is accessed directly there will be no book to find - so go fetch
  useEffect(() => {
    if (books.length < 1) {
      dispatch(loadBooks());
    }
  }, [dispatch, books]);

  let render = books.length ? <BookResults books={books}/> : <Loading />;

  return (
    <React.Fragment>
      {render}
    </React.Fragment>
  );
};

Search.propTypes = {
  books: PropTypes.arrayOf(PropTypes.object).isRequired,
};

const mapStateToProps = (state) => {
  return {
    books: state.books,
  };
};

export default Layout(connect(mapStateToProps)(Search));
