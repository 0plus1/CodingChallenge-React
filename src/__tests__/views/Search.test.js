import React from 'react';
import Search from '../../views/Search';
import { shallowWithRouter } from '../../testHelper';
import mockAxios from 'axios';
import bookApi from '../../api/books';

describe('<Search />', () => {
  describe('renders', () => {
    it('without crashing', async () => {
      mockAxios.get.mockImplementationOnce(() => Promise.resolve({
          data: [
            {
              book_id: 2,
              name: 'Schuster Champlinborough',
              isbn: '1441614311',
              published_at: '1991-07-29',
              author: 'Prof. John Doe',
              cover: 'https:\/\/lorempixel.com\/640\/480\/?84662'
            }
          ]
        })
      );

      shallowWithRouter(<Search/>);
      expect(<Search/>)
        .toMatchSnapshot();

      await bookApi.getBooks();

      expect(mockAxios.get)
        .toHaveBeenCalledTimes(1);
    });
  });
});
