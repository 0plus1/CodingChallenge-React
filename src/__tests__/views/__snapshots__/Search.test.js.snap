// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders without crashing 1`] = `<PureComponent />`;
